test/framework/src/main/java/org/elasticsearch/test/XContentTestUtils.java

```diff
@@ -123,4 +136,163 @@ private static String differenceBetweenObjectsIgnoringArrayOrder(String path, Ob
+            };
+            return XContentTestUtils
+                    .insertIntoXContent(contentType.xContent(), xContent, insertPaths, () -> randomAsciiOfLength(random, 10), value)
+                    .bytes();
+        }
+    }
+
+    /**
+     * This utility method takes an XContentParser and walks the xContent structure to find all
+     * possible paths to where a new object or array starts. This can be used in tests that add random
+     * xContent values to test parsing code for errors or to check their robustness against new fields.
+     *
+     * The path uses dot separated fieldnames and numbers for array indices, similar to what we do in
+     * {@link ObjectPath}.
+     *
+     * As an example, the following json xContent:
+     * <pre>
+     *     {
+     *         "foo" : "bar",
+     *         "foo1" : [ 1, { "foo2" : "baz" }, 3, 4]
+     *         "foo3" : {
+     *             "foo4" : {
+     *                  "foo5": "buzz"
+     *             }
+     *         }
+     *     }
+     * </pre>
+     *
+     * Would return the following list:
+     *
+     * <ul>
+     *  <li>"" (the empty string is the path to the root object)</li>
+     *  <li>"foo1.1"</li>
+     *  <li>"foo3</li>
+     *  <li>"foo3.foo4</li>
+     * </ul>
+     */
+    static List<String> getInsertPaths(XContentParser parser) throws IOException{
+        assert parser.currentToken() == null : "the parsers token cursor should be before the first token";
+        assert parser.nextToken() == XContentParser.Token.START_OBJECT;
+        return getInsertPaths(parser, new Stack<String>());
+    }
+
+    private static List<String> getInsertPaths(XContentParser parser, Stack<String> currentPath) throws IOException {
+        assert parser.currentToken() == XContentParser.Token.START_OBJECT || parser.currentToken() == XContentParser.Token.START_ARRAY :
+            "should only be called when new objects or arrays start";
+        List<String> validPaths = new ArrayList<>();
+        // parser.currentName() can be null for root object and unnamed objects in arrays
+        if (parser.currentName() != null) {
+            currentPath.push(parser.currentName());
+        }
+        if (parser.currentToken() == XContentParser.Token.START_OBJECT) {
+            validPaths.add(String.join(".", currentPath.toArray(new String[currentPath.size()])));
+            while (parser.nextToken() != XContentParser.Token.END_OBJECT) {
+                if (parser.currentToken() == XContentParser.Token.START_OBJECT
+                        || parser.currentToken() == XContentParser.Token.START_ARRAY) {
+                    validPaths.addAll(getInsertPaths(parser, currentPath));
+                }
+            }
+        } else if (parser.currentToken() == XContentParser.Token.START_ARRAY) {
+            int itemCount = 0;
+            while (parser.nextToken() != XContentParser.Token.END_ARRAY) {
+                if (parser.currentToken() == XContentParser.Token.START_OBJECT
```

Comment:
not sure if worth it, but maybe we could have methods that specifically handle arrays and objects, which can be called directly rather than having to check again the current token. Maybe this complicates things too much. It may help also with the currentName null check, as it can be moved to only the array method.